File: frontend/src/lib/api/ai-usage.ts
/**
 * API client functions for AI usage tracking.
 */
import { api } from './client';

// =============================================================================
// Types
// =============================================================================

export interface AIUsageRecord {
  id: string;
  student_id: string;
  date: string;
  tokens_haiku: number;
  tokens_sonnet: number;
  total_tokens: number;
  total_cost_usd: string;
  request_count: number;
  created_at: string;
  updated_at: string;
}

export interface AIUsageLimits {
  today_tokens: number;
  today_cost_usd: string;
  today_requests: number;
  month_tokens: number;
  month_cost_usd: string;
  daily_token_limit: number;
  monthly_soft_limit: number;
  monthly_hard_limit: number;
  daily_limit_reached: boolean;
  monthly_soft_limit_reached: boolean;
  monthly_hard_limit_reached: boolean;
  daily_usage_percent: number;
  monthly_usage_percent: number;
}

export interface AIUsageSummary {
  student_id: string;
  period_start: string;
  period_end: string;
  total_tokens_haiku: number;
  total_tokens_sonnet: number;
  total_tokens: number;
  total_cost_usd: string;
  total_requests: number;
  daily_average_tokens: number;
}

export interface AIUsageHistoryResponse {
  usage: AIUsageRecord[];
  total: number;
  page: number;
  page_size: number;
}

// =============================================================================
// API Functions
// =============================================================================

/**
 * Get current AI usage limits and status for a student.
 */
export async function getAIUsageLimits(studentId: string): Promise<AIUsageLimits> {
  const response = await api.get<AIUsageLimits>(
    `/students/${studentId}/ai-usage/limits`
  );
  return response;
}

/**
 * Get AI usage history for a student.
 */
export async function getAIUsageHistory(
  studentId: string,
  days: number = 30
): Promise<AIUsageRecord[]> {
  const response = await api.get<AIUsageRecord[]>(
    `/students/${studentId}/ai-usage/history`,
    { params: { days: days.toString() } }
  );
  return response;
}

/**
 * Get AI usage summary for a student over a period.
 */
export async function getAIUsageSummary(
  studentId: string,
  startDate: string,
  endDate: string
): Promise<AIUsageSummary> {
  const response = await api.get<AIUsageSummary>(
    `/students/${studentId}/ai-usage/summary`,
    { params: { start_date: startDate, end_date: endDate } }
  );
  return response;
}

/**
 * Check if a student can make an AI request.
 */
export async function canMakeAIRequest(
  studentId: string,
  estimatedTokens: number = 0
): Promise<{ allowed: boolean; reason?: string }> {
  const response = await api.get<{ allowed: boolean; reason?: string }>(
    `/students/${studentId}/ai-usage/can-request`,
    { params: { estimated_tokens: estimatedTokens.toString() } }
  );
  return response;
}

File: backend/app/schemas/ai_usage.py
"""AI usage tracking schemas."""
from datetime import date, datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, Field


# =============================================================================
# Response Schemas
# =============================================================================


class AIUsageResponse(BaseModel):
    """Daily AI usage record."""

    id: UUID
    student_id: UUID
    date: date
    tokens_haiku: int
    tokens_sonnet: int
    total_tokens: int
    total_cost_usd: Decimal
    request_count: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class AIUsageSummary(BaseModel):
    """Summary of AI usage for a period."""

    student_id: UUID
    period_start: date
    period_end: date
    total_tokens_haiku: int = 0
    total_tokens_sonnet: int = 0
    total_tokens: int = 0
    total_cost_usd: Decimal = Decimal("0")
    total_requests: int = 0
    daily_average_tokens: int = 0


class AIUsageLimits(BaseModel):
    """Current usage limits and status."""

    # Current usage (today)
    today_tokens: int = 0
    today_cost_usd: Decimal = Decimal("0")
    today_requests: int = 0

    # Monthly usage
    month_tokens: int = 0
    month_cost_usd: Decimal = Decimal("0")

    # Limits
    daily_token_limit: int = Field(default=150_000)
    monthly_soft_limit: int = Field(default=2_000_000)
    monthly_hard_limit: int = Field(default=3_000_000)

    # Status
    daily_limit_reached: bool = False
    monthly_soft_limit_reached: bool = False
    monthly_hard_limit_reached: bool = False

    # Percentage used
    daily_usage_percent: float = 0.0
    monthly_usage_percent: float = 0.0


class AIUsageHistoryResponse(BaseModel):
    """Paginated AI usage history."""

    usage: list[AIUsageResponse]
    total: int
    page: int
    page_size: int


# =============================================================================
# Internal Schemas
# =============================================================================


class AIUsageUpdate(BaseModel):
    """Internal schema for updating AI usage."""

    tokens_haiku: int = 0
    tokens_sonnet: int = 0
    cost_usd: Decimal = Decimal("0")
    requests: int = 1
